_posts/2015-05-11-traits.md

```diff
@@ -0,0 +1,420 @@
+code execution and data representation.
+
+This post will walk through each of the above points at a high level, to give
+you a sense for how the design achieves these goals, without getting too bogged
+down in the details.
+
+### Background: methods in Rust
+
+> Before delving into traits, we need to look at a small but important detail of
+> the language: the difference between methods and functions.
+
+Rust offers both methods and free-standing functions, which are very
+closely related:
+
+```rust
+struct Point {
+    x: f64;
+    y: f64,
+}
+
+// a free-standing function that converts a (borrowed) point to a string
+fn point_to_string(point: &Point) -> String { ... }
+
+// an "inherent impl" block defines the methods available directly on a type
+impl Point {
+    // this method is available on any Point, and automatically borrows the
+    // Point value
+    fn to_string(&self) -> String { ... }
+}
+```
+
+Methods like `to_string` above are called "inherent" methods, because they:
+
+* Are tied to a single concrete "self" type (specified via the `impl` block header).
+* Are *automatically* available on any value of that type -- that is, unlike
+  functions, inherent methods are always "in scope".
+
+The first parameter for a method is always an explicit "self", which is either
+`self`, `&mut self`, or `&self` depending on the
+[level of ownership required][socket].  Methods are invoked using the `.`
+notation familiar from object-oriented programming, and the self parameter is
+*implicitly borrowed* as per the form of `self` used in the method:
+
+```rust
+let p = Point { x: 1.2, y: -3.7 };
+let s1 = point_to_string(&p);  // calling a free function, explicit borrow
+let s2 = p.to_string();        // calling a method, implicit borrow as &p
+```
+
+Methods and their auto-borrowing are an important aspect of the ergonomics of
+Rust, supporting "fluent" APIs like the one for spawning processes:
+
+```rust
+let child = Command::new("/bin/cat")
+    .arg("rusty-ideas.txt")
+    .current_dir("/Users/aturon")
+    .stdout(Stdio::piped())
+    .spawn();
+```
+
+### Traits are interfaces
+
+Interfaces specify the expectations that one piece of code has on another,
+allowing each to be switched out independently. For traits, this specification
+largely revolves around methods.
+
+Take, for example, the following simple trait for hashing:
+
+```rust
+trait Hash {
+    fn hash(&self) -> u64;
+}
+```
+
+In order to implement this trait for a given type, you must provide a `hash`
+method with matching signature:
+
+```rust
+impl Hash for bool {
+    fn hash(&self) -> u64 {
+        if *self { 0 } else { 1 }
+    }
+}
+
+impl Hash for i64 {
+    fn hash(&self) -> u64 {
+        *self as u64
+    }
+}
+```
+
+Unlike interfaces in languages like Java, C# or Scala, new traits can be implemented
+for existing types (as with `Hash` above). **Code in upstream crates can be
+adapted to interfaces in downstream crates**.
```

Comment:
This sentence threw me for a bit of a spin as I was trying to figure out what was up and what was down, and I was wondering if perhaps the "new traits can be implemented for existing types" can be the emphasis here with the implication that you can add new methods and enhance functionality? (or just a minor rewording would probably do it!)